aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp

```diff
@@ -0,0 +1,556 @@
+#include <ATen/ATen.h>
+#include <ATen/NestedTensorImpl.h>
+#include <ATen/native/nested/cuda/NestedTensorTransformerUtils.h>
+#include <iostream>
+#include <tuple>
+namespace at {
+namespace native {
+namespace preprocessing {
+
+namespace {
+
+/**
+ * This builds up the cumulative sequence length for a batch of sequences.
+ * This is not very dry, but in the backward pass we already have cumulative_seq_len
+ * on device. And all we need on CPU to launch the kernel is NNz. We could refactor the
+ * the below function but it adds more complexity than I think is needed.
+ */
+int64_t get_nnz(Tensor nestedtensor) {
+  auto* nt_impl = get_nested_tensor_impl(nestedtensor);
+  const auto& sizes = nt_impl->get_nested_sizes();
+  auto size_tensor_stride = sizes.stride(0);
+  const int64_t batch_size = nestedtensor.size(0);
+  auto* sizes_ptr = sizes.data_ptr<int64_t>();
+  int64_t cumulative_sequence_length = 0;
+  for (const auto i : c10::irange(batch_size)) {
+    // Calculate the cumulative sum of the sequence lengths
```

Comment:
I could also use storage_offsets with some math to divide out by the n_heads and head_dim... not sure though if it is worth it/ makes it a little more abstract